File: backend/app/schemas/wallet.py
from pydantic import BaseModel
from datetime import datetime


class WalletResponse(BaseModel):
    balance: int
    locked_balance: int
    total: int
    can_rescue: bool = False


class LedgerEntryResponse(BaseModel):
    id: str
    type: str
    amount: int
    balance_after: int
    reference_id: str | None
    description: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class PaginatedLedgerResponse(BaseModel):
    items: list[LedgerEntryResponse]
    total: int
    limit: int
    offset: int


class RescueResponse(BaseModel):
    balance: int
    ledger_entry: LedgerEntryResponse

File: frontend/src/api/wallet.ts
import api from './client'

export interface WalletInfo {
  balance: number
  locked_balance: number
  total: number
  can_rescue: boolean
}

export interface LedgerEntry {
  id: string
  type: string
  amount: number
  balance_after: number
  description: string | null
  created_at: string
}

export interface LedgerResponse {
  items: LedgerEntry[]
  total: number
  pages: number
  page: number
}

const PAGE_SIZE = 20

export const walletApi = {
  get: () =>
    api.get<WalletInfo>('/wallet').then(r => ({
      ...r,
      data: { ...r.data, can_rescue: r.data.can_rescue ?? false },
    })),

  ledger: (page = 1) =>
    api.get<{ items: LedgerEntry[]; total: number }>(`/wallet/ledger?offset=${(page - 1) * PAGE_SIZE}&limit=${PAGE_SIZE}`).then(r => ({
      ...r,
      data: {
        items: r.data.items,
        total: r.data.total,
        pages: Math.max(1, Math.ceil(r.data.total / PAGE_SIZE)),
        page,
      } as LedgerResponse,
    })),

  rescue: () =>
    api.post<WalletInfo>('/wallet/rescue'),
}
